src/Features/LanguageServer/Protocol/Handler/Formatting/FormatDocumentOnTypeHandler.cs

```diff
@@ -19,12 +19,10 @@
 
 namespace Microsoft.CodeAnalysis.LanguageServer.Handler
 {
-    [ExportRoslynLanguagesLspRequestHandlerProvider, Shared]
-    [ProvidesMethod(Methods.TextDocumentOnTypeFormattingName)]
+    [ExportRoslynLanguagesLspRequestHandlerProvider(typeof(FormatDocumentOnTypeHandler)), Shared]
+    [Method(Methods.TextDocumentOnTypeFormattingName)]
     internal class FormatDocumentOnTypeHandler : AbstractStatelessRequestHandler<DocumentOnTypeFormattingParams, TextEdit[]?>
```

Comment:
So something that threw me off while reviewing this: we have lots of exports that are exporting a HandlerProvider, but the base type here implies we're inheriting a Handler. But I see that base type actually implementing both the handler and also provider interfaces simultaneously. Doesn't have to be a part of this PR, but what might be helpful would be to do something like:

1. Rename this to "AbstractStatelessRequestHandlerProvider".
2. Delete the fact this base type actually implements IRequestHandler; instead just have an abstract method with the same signature.
3. Have the CreateRequestHandlers call just make a thin shim object that forwards.

Otherwise it's really hard to get a sense of whether or not these are mismatching exports and interfaces or not. :-/

Or maybe just call it AbstractStatelessRequestHandlerAndProvider so it's clear it's implementing both?

Reply:
updated in https://github.com/dotnet/roslyn/pull/59600